pages/search/search.jsx

```diff
@@ -51,24 +50,97 @@ export default React.createClass({
       location[`query`] = { keyword: query };
     }
 
+    // note browserHistory.push() triggers component re-render
     browserHistory.push(location);
+  },
+  clearSearch() {
+    this.refs.searchInput.setState({value: ``}, () => {
+      this.updateBrowserHistory();
+    });
+  },
+  fetchData(params = {}, entries = [], apiPageIndex = 1) {
+    if (!params.search) {
+      // if no search keyword was passed reset search page back to its initial state
+      this.setState(this.getInitialState());
+      return;
+    }
+
+    params.page = apiPageIndex;
+
+    Service.entries
+      .get(params)
+      .then((data) => {
+        entries = entries.concat(data.results);
 
+        if (data.next) {
+          // there are more matched entries in the database we need to fetch
+          this.fetchData(params, entries, apiPageIndex+1);
+        } else {
+          // Multiple XHRs to Pulse API can happen around the same time and
+          // we can't guarantee which one we hear back first.
+          // However, we only care about capturing entries that are matched with
+          // our current search keyword. If entries returned from Pulse API here
+          // belong to any of our previous search keywords, we can ignore them by
+          // doing nothing.
+          if (params.search !== this.state.keywordBeingSearched) {
+            return;
+          }
+          this.setState({
+            keywordBeingSearched: ``,
+            keywordSearched: params.search,
+            entriesMatched: entries
+          });
+        }
+      })
+      .catch((reason) => {
+        console.error(reason);
+      });
+  },
+  handleInputChange() {
+    this.updateBrowserHistory();
     this.setState({
-      searchQuery: query
+      keywordBeingSearched: this.refs.searchInput.state.value
     });
+    this.fetchData({search: this.refs.searchInput.state.value});
+  },
+  handleDismissBtnClick() {
+    this.clearSearch();
   },
   render() {
+    let projects;
+    let showViewMoreBtn;
+    let searchResult;
+
+    if (this.state.keywordSearched) {
+      let numEntriesMatched = this.state.entriesMatched.length;
```

Comment:
If we don't fetch all the entries at the start, we can just use the `count` property on the API response here